pep-0561.rst

```diff
@@ -123,8 +123,10 @@ For namespace packages (see PEP 420), the ``py.typed`` file should be in the
 submodules of the namespace, to avoid conflicts and for clarity.
 
 This PEP does not support distributing typing information as part of
-module-only distributions. The code should be refactored into a package-based
-distribution and indicate that the package supports typing as described
+module-only distributions or single-file modules within namespace packages.
+
+The single-file module should be refactored into a package
+and indicate that the package supports typing as described
```

Comment:
I don't think is quite right. In the above issue, the `py.typed` file belongs inside the `google/cloud` folder, as `google` is the namespace package (unless I am mistaken), so the `py.typed` file lives in each sub-package of the namespace package.

Reply:
google.cloud is also a namespace package (sorry for confusion)

Reply:
It's a nested namespace package

Reply:
I believe you should still list the sub-package of the outermost package as typed (even if partial).

Reply:
Just to make sure we're on the same page - I don't think the preexisting PEP unambiguously specifies this case (nested namespace packages) - we are discussing/reasoning about what makes the most sense to make as the specification here.

The sub-namespace-package (`google.cloud` in this case) will be populated by several different pip distributions (https://pypi.org/search/?q=google-cloud-). I don't think it works - or makes sense - for each of those pip distributions to place a `py.typed partial` in `google.cloud`.

As a concrete example - for something like `google/cloud/firestore.py` which is fully typed - It doesn't make sense to have a `google/cloud/py.typed` indicating full typed-ness - since `google/cloud` could have other entries from other distributions. Rather we must require it to be `google/cloud/firestore/__init__.py` with `google/cloud/firestore/py.typed` so that typecheckers can know that `firestore` is fully typed.

Reply:
Yeah that seems reasonable.

Reply:
@nipunn1313 I ran into something related to this while working on typing for https://pypi.org/project/googleapis-common-protos/ and was wondering if you could confirm if my understanding of the updated spec is correct.

Like most Google distributions this distribution contains a namespace `google` package, which contains multiple subpackages. Here's the tree showing one particular subpackage:
```
google
├── __init__.py
├── rpc
│   ├── code_pb2.py
│   ├── code.proto
│   ├── context
│   │   ├── attribute_context_pb2.py
│   │   ├── attribute_context.proto
│   │   └── __init__.py
│   ├── error_details_pb2.py
│   ├── error_details.proto
│   ├── __init__.py
│   ├── README.md
│   ├── status_pb2.py
│   └── status.proto
```
Here _both_ `google` and `rpc` are namespace packages (using `declare_namespace`/`extend_path`), while `context` is a regular package. As you can see `google.rpc` actually contains some modules. Presumably some other Google distribution could put other modules in this namespace. If I understand this discussion correctly, there's no way to distribute typing information for `code_pb2.py`, `error_details_pb2.py` and `status_pb2.py` and for this to work Google would have to refactor these into packages? Is the correct approach when making a stub-only package for this to simply leave them out, like this?
```
google-stubs
├── rpc
│   └── context
│       ├── attribute_context_pb2.pyi
│       ├── __init__.pyi
|       └── py.typed
```

Reply:
I actually wrote this spec modification with a case in mind - single-file-modules where types are included in the original package (googleapis-common-protos in your eg) - because in that case - a py.typed is required.

For a stub-only package - By the spec clarification - all namespace packages within stub-packages are to be considered partially typed by type checkers.
```
Type checkers should treat namespace packages within stub-packages as
incomplete since multiple distributions may populate them.
Regular packages within namespace packages in stub-package distributions
are considered complete unless a ``py.typed`` with ``partial\n`` is included.
```

Thus I think you can opt to do
```
google-stubs
├── rpc
│   └── context
│       ├── attribute_context_pb2.pyi
│       ├── __init__.pyi
|       └── py.typed
```
You may also add stubs for the remaining files (eg `status_pb2.pyi`), but be sure to omit the `__init__.pyi` for namespace packages. Typechecker should consider `google-stubs` and `google-stubs/rpc` as incomplete because they are namespace packages.


I haven't tested whether mypy/pyright obey this spec clarification yet - but I believe pyright recently made a [change](https://github.com/microsoft/pyright/issues/2113#issuecomment-932688627) to make this work. I believe in my past testing mypy needed the `--namespace-packages` flag to treat these properly, but I didn't test this specific case.